Adds netlink to navstar overlay

GoelDeepak opened the pull request.

GoelDeepak committed:
Adds netlink to navstar overlay

GoelDeepak committed:
Moves configurator to netlink

GoelDeepak committed:
Minor test cleanup and gen_netlink ref update

GoelDeepak commented:
@sargun ... This is ready for review

GoelDeepak committed:
removes unused code

sargun commented on apps/navstar_overlay/src/navstar_overlay_netlink.erl:
```diff
@@ -0,0 +1,170 @@
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-define(TEST_AND_DEV, true).
```
TEST_OR_DEV?

sargun commented on apps/navstar_overlay/src/navstar_overlay_netlink.erl:
```diff
@@ -0,0 +1,170 @@
+  Attr = [{dst, Dst}, {lladdr, Lladdr}],
+  Neigh = {
+    _Family = inet,
+    _Ifindex = if_nametoindex(Ifname),
```
Please put this on a separate line.

sargun replied:
In general, if you're doing something that has side-effects while initializing a Tuple, it's a bit difficult to track down when it breaks.

sargun commented:
Two things:
1) It would be good to add some tests
2) It would be good to have an "AAE" mode that runs every once in a while which makes sure all the neighbor and routes are matching as we expect them to be (non-delta config application)

sargun commented:
otherwise, looks good.

GoelDeepak committed:
minor code refactoring

GoelDeepak committed:
Adds reapplying state to reapply the config periodically

GoelDeepak committed:
Adds circle ci tests for navstar overlay netlink

GoelDeepak committed:
redesign the state machine

sargun commented on apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl:
```diff
@@ -18,22 +18,24 @@
 
--type config() :: #{OverlayKey :: term() := Value :: term()}.
--type config2() :: #{key := term(), value := term()}.
+-type orddict() :: [{Key::term(), Value::term()}].
```
Any reason you're not using orddict:ordict(Type1, Type1)?

sargun commented on apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl:
```diff
@@ -18,22 +18,24 @@
--type config() :: #{OverlayKey :: term() := Value :: term()}.
--type config2() :: #{key := term(), value := term()}.
+-type orddict() :: [{Key::term(), Value::term()}].
+-type config() :: orddict().
```
You should probably use orddict:orddict and specify the key and val types

sargun commented on apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl:
```diff
@@ -114,68 +116,121 @@ terminate(_Reason, _State, _Data) ->
-                      ordsets:union(ordsets:from_list(DeltaValue), OldValue)
-               end,
-    maps:put(OverlayKey, NewValue, OldConfig).
+-spec(handle_event(LashupEvent :: tuple(), #data{}) -> #data{}).
```
Don't use handle_event, it's a special function name when gen_statem is in a different callback mode.

sargun commented on apps/navstar_overlay/src/navstar_overlay_lashup_kv_listener.erl:
```diff
@@ -114,68 +116,121 @@ terminate(_Reason, _State, _Data) ->
-               end,
-    maps:put(OverlayKey, NewValue, OldConfig).
+-spec(handle_event(LashupEvent :: tuple(), #data{}) -> #data{}).
+handle_event({lashup_kv_events, #{type := ingest_new, key := OverlayKey, value := NewValue, ref := Ref}},
```
Can we come up with a better name than "Value" Maybe OverlayConfiguration?

sargun commented on rebar.config:
```diff
@@ -18,10 +18,11 @@
 ]}.
 
-{relx, [{release, {'navstar', "0.1.0"}, [navstar, mnesia, eper, runtime_tools, eflame, debugger, observer]},
+{relx, [{release, {'navstar', "0.1.0"}, [navstar, mnesia, eper, runtime_tools, eflame, debugger, observer, {gen_netlink, load}]},
```
Why `{gen_netlink, load}` It's okay to start it.

GoelDeepak committed:
Fixes circle ci issue

with sudo, circle ci has issues in downloading some of the dependencies.
For example, lint. To fix this issue we need to do 'make check' without
sudo so that all the dependencies are downloaded and then run with
sudo to run ct tests

GoelDeepak committed:
Use lashup keys(MatchSpec) to fetch navstar overlay keys

GoelDeepak committed:
adds gen_netlink dependency
